Documentation/wip/cli-extensibility.md

```diff
@@ -0,0 +1,128 @@
+     environment, env              Manage environments
+     watch, w                      Manage watches
+     package, pkg                  Manage packages
+     spec, specs                   Manage a declarative app specification
+     upgrade                       Upgrade tool from fission v0.1
+     tpr2crd                       Migrate tool for TPR to CRD
+     help, h                       Shows a list of commands or help for one command
+
+PLUGINS:
+     workflows, wf                 Inspect and manage workflow executions
+     ui                            Start the user interface
+
+GLOBAL OPTIONS:
+   --server value  Fission server URL (default: "http://127.0.0.1:65356")
+   --help, -h      show help
+   --version, -v   print the version
+```
+Of course Fission needs to be able to find all plugins for this. There are several ways in which we can provide discoverability. The simplest one is for Fission to look in the path for all binaries starting with the `fission-*` prefix. Optionally, fission could invoke a specific command on the subcommand to get info about the plugin (such as version, help text, aliases...)
+
+With Fission Workflows this info would look something like this: 
+```bash
+$ fission-workflows --plugin
+name: workflows
+version: 0.4.0
+help: Inspect and manage workflow executions
+```
+The idea is that this plugin info is all completely optional. 
+If it is not available, we simply degrade the results to user. 
+This way users/we can easily prototype or add plugins without having to worry about adhering to some interface.
+
+**List version**
+```bash
+$ fission --version
+client: 
+  fission: 0.8.0
+  fission-workflows: 0.4.0
+server:
+  fission: 0.8.1
+  fission-workflows: 0.3.0
+```
+Again, versioning info for fission-workflows is taken from the plugin info of the commands. 
+Note: a related issue is to have some more formalized plugin support/discoverability on the server-side, 
+but that is out of the scope of this issue.
+
+### Other (optional) extensions and notes
+- Like git we could setup a preferred binary path, where fission looks first when searching for the subcommand. 
+This could optionally be defined with a `FISSION_EXEC_PATH`.
+- With the current approach we cannot have aliases for commands---fission will not be able to find fission-workflows 
+when the user calls `fission wf`. This might be UX issue, with these long path names. One option is let the user fix 
+it themselves by symlinking `fission-wf` to `fission-workflows`; using the plugin info Fission can recognize and 
+merge aliases together.
+- To help detect versioning conflicts (old version of fission, too new version of fission workflows). We could add 
+a `requires` field to the fission-workflows plugin info. Then we could throw a warning or error, when two out of sync 
+versions are being used. 
+- To avoid unhelpful errors to the user when they have not installed a plugin, we could add a heuristic to check 
```

Comment:
I think we should compile-in a list of known plugins and URLs for where to download them, and let the user handle the rest for now.  For unknown plugins you can output the generic text.  Later we can get fancier and fetch the plugin list from a server, download and install plugins, etc.

Reply:
Like the previous comment, this part is a bit outdated. The current flow:
1. Check if command is present locally; if so, execute it as a subcommand.
2. Use a hard-coded url (currently `https://github.com/fission/fission/blob/master/plugins.yaml`) to fetch plugin suggestions. If the requested command is in there, print a suggestion how to download and install it.
3. Otherwise, print the current "command %v not found" warning.